org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend

```diff
@@ -20,21 +22,31 @@ import org.eclipse.emf.common.util.URI
 class UriExtensions {
 
 	def URI toUri(String pathWithScheme) {
-		val javaNetUri = java.net.URI.create(pathWithScheme);
-		val path = javaNetUri.toPath
-		return URI.createURI(path)
+		// URI is used to get path of the uri without scheme
+		var path = URI.createURI(pathWithScheme).path
+		return URI.createURI(path.toPath)
 	}
 
 	def String toPath(URI uri) {
-		return java.net.URI.create(uri.toString).toPath
+		return URI.createURI(uri.path.toPath).toString
 	}
 
 	def String toPath(java.net.URI uri) {
+		return uri.path.toPath
+	}
+
+	/**
+	 * We need to check if current path represents directory in file system
+	 * and need to add trailing slash if path represents directory.
```

Comment:
Is the comment up to date? If yes, could you please add a testcase? Or did I just miss something?

Reply:
There is a test for that, UriExtensionsTest.testFolderIsPrefix.

Reply:
That relation was not obvious to me. Sorry for the unnecessary noise.

Reply:
Could you please clarify that, why this is required? Assuming a very simple test case, the `isPrefix` returns with a `false` for the `URI.createFileURI`. So basically, when I have an EMF URI and I convert it to string and back to URI then I would get a different URI than I had initially. (Or I have just overlooked something which could easily happen).

Test case:
```java
	@Test
 	def void testFolderIsPrefix_why() {
 		var directory = new File("./test-data/test-project")
 		assertTrue(URI.createFileURI(directory.absolutePath).isPrefix)
 	}
```

(Sorry for the formatting.)

Reply:
Because `UriUtil.isPrefixOf` checks if `URI` has trailing path separator. It is used in `FileProjectConfig`. If we do not add trailing separator `FileProjectConfig` will not find any projects associated with given file. That is why I convert input URI to string path and create a new hierarchical one. I did not realize that it removes `inmemory:/` scheme, since it passed tests with this scheme.

Reply:
Cool, thanks for the update! One more question, if I have this:

```
var directory = new File("./test-data/test-project")
URI.createFileURI(directory.absolutePath).isPrefix //false
```

Why would you expect to have `isPrefix` `true` after the string then to URI conversion. Besides that, this is a behavior change if I understood you correctly, and it is unrelated to the URL encoding/decoding issue at all, right?

PS: and do not feel bad because of the `inmemory:/` problem, that's why we need to add test and do the review correctly.

Reply:
> and it is unrelated to the URL encoding/decoding

If that is true, can we handle this in another PR/issue?

Reply:
`File("./test-data/test-project")` is an actual directory, located in `org.eclipse.xtext.ide.tests` during tests. It calls `toPath(String)` method, that uses `Paths.get()` to form a path and converts it to URI. And `java.nio.file.Path.toUri()` states that
> ... if file is a directory, then the resulting URI will end with a slash

So it should return true it that case. Maybe it is not the best idea, since it depends on other tests.
Original Cyrillic letter decoding is fixed by `URLDecoder`, but while debugging I found out that spaces are not supported either, because `java.net.uri` does not support space symbol. I`m not sure if it is related to URI decoding in general, but yes, it is not related to Cyrillic letters decoding.

Reply:
This conversion shouldn't be done in general, but rather where needed. Also the code you mentioned seems to take care of it already : https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext/src/org/eclipse/xtext/workspace/FileProjectConfig.xtend#L98

Reply:
Could you tell us the concrete scenario in which the `isPrefixOf` didn't return the desired result?